tests/test_barostat.py

```diff
@@ -0,0 +1,155 @@
+    initial_waterbox_width = 2.0 * unit.nanometer
+    timestep = 1.5 * unit.femtosecond
+    collision_rate = 1.0 / unit.picosecond
+    n_moves = 2000
+    barostat_interval = 5
+    seed = 2021
+
+    # thermodynamic parameters
+    temperatures = np.array([300, 600, 1000]) * unit.kelvin
+    pressure = 1.013 * unit.bar
+
+    # generate an alchemical system of a waterbox + alchemical ligand:
+    # effectively discard ligands by running in AbsoluteFreeEnergy mode at lambda = 1.0
+    mol_a, _, core, ff = hif2a_ligand_pair.mol_a, hif2a_ligand_pair.mol_b, hif2a_ligand_pair.core, hif2a_ligand_pair.ff
+    complex_system, complex_coords, complex_box, complex_top = build_water_system(
+        initial_waterbox_width.value_in_unit(unit.nanometer))
+
+    min_complex_coords = minimize_host_4d([mol_a], complex_system, complex_coords, ff, complex_box)
+    afe = AbsoluteFreeEnergy(mol_a, ff)
+
+    _unbound_potentials, _sys_params, masses, coords = afe.prepare_host_edge(
+        ff.get_ordered_params(), complex_system, min_complex_coords
+    )
+
+    # drop the nonbonded potential
+    unbound_potentials = _unbound_potentials[:-1]
+    sys_params = _sys_params[:-1]
+
+    # get list of molecules for barostat by looking at bond table
+    harmonic_bond_potential = unbound_potentials[0]
+    bond_list = get_bond_list(harmonic_bond_potential)
+    group_indices = get_group_indices(bond_list)
+
+    trajs = []
+    volume_trajs = []
+
+    potential_energy_model = PotentialEnergyModel(sys_params, unbound_potentials)
+    lam = 1.0
+
+    n_molecules = complex_top.getNumResidues()
+
+    # expected volume
+    md_pressure_unit = ENERGY_UNIT / DISTANCE_UNIT ** 3
+    pressure_in_md = (pressure * unit.AVOGADRO_CONSTANT_NA).value_in_unit(md_pressure_unit)
+    expected_volume_in_md = (n_molecules + 1) * BOLTZ * temperatures.value_in_unit(unit.kelvin) / pressure_in_md
+
+
+    for i, temperature in enumerate(temperatures):
+        # define NPT ensemble
+        ensemble = NPTEnsemble(potential_energy_model, temperature, pressure)
+
+        # define a thermostat
+        integrator = LangevinIntegrator(
+            temperature.value_in_unit(unit.kelvin),
+            timestep.value_in_unit(unit.picosecond),
+            collision_rate.value_in_unit(unit.picosecond**-1),
+            masses,
+            seed
+        )
+        integrator_impl = integrator.impl()
+
+        def reduced_potential_fxn(x, box, lam):
+            u, du_dx = ensemble.reduced_potential_and_gradient(x, box, lam)
+            return u
+
+        thermostat = UnadjustedLangevinMove(integrator_impl, potential_energy_model.all_impls, lam, n_steps=barostat_interval)
+        barostat = MonteCarloBarostat(partial(reduced_potential_fxn, lam=lam), group_indices, max_delta_volume=3.0)
+
+        v_0 = sample_velocities(masses * unit.amu, temperature)
+
+        # rescale the box to be approximately the desired box volume already
+        rescaler = CentroidRescaler(group_indices)
```

Comment:
If you disable the barostat in this test, this test still passes

Reply:
Explainable (but undesirable!) behavior, since the test is initialized at the expected volume here https://github.com/proteneer/timemachine/pull/433/commits/5f9cf6dccd2021060ec6033c260cc346dc0adc71 .

The initial version of the test took a long time to run, since it required to equilibrate to a far larger box size than initialized (like, initialized at ~10-100 nm^3, with ~10000 nm^3 expected). However, changing the initialization to be _exactly_ the expected box volume has the undesirable effect of making the test only check whether running the barostat causes the simulation to _drift away from the correct answer._ We of course also would want this test to fail if the barostat is doing nothing!

The behavior of the test should be fixed in https://github.com/proteneer/timemachine/pull/433/commits/05817c851e676ccb40042ca2b121adbfa367213b , where the barostat is initialized from a box that's instead 10% too large.

Reply:
And with this aspect of the test behavior fixed, it appears the test is no longer passing!

This needs to be resolved before merging.

Possibilities:
* there are uncaught bugs with the barostat itself
* the test is not behaving as intended for another reason, e.g. because simulation with such a giant box is causing other numerical issues or slow thermostat convergence. (Note that the OpenMM version of this test system has only 64 particles, and much smaller expected box sizes...)

Reply:
[Increasing the pressure by a factor of 100](https://github.com/proteneer/timemachine/pull/433/commits/4886d6ec176d26d70a8bdc2ee70aa9e951e89781) so that the expected volume is within a couple orders of magnitude of water box simulation cells we typically consider, the test is failing but with small relative differences from expectation.

```
Mismatched elements: 1 / 3 (33.3%)
Max absolute difference: 4.03597219
Max relative difference: 0.01881098
 x: array([107.001069, 218.590044, 360.868921])
 y: array([107.277036, 214.554072, 357.59012 ])
```

Will adjust to run this test longer...

Reply:
Adjusting the molecular ideal gas tests to [run 5x longer](https://github.com/proteneer/timemachine/pull/433/commits/ba0e6676d8198b23244425e0c52a73963e207938) causes this to pass, reducing my level of suspicion of an uncaught bug in the barostat itself. However, this adds another 3 minutes to the tests...

Reply:
Could limiting `max_delta_volume=3.0` here be a source of slow convergence? An optimal choice (or adaptively identified choice) would presumably converge pretty quickly if you're just testing the mean.

You can also cut your convergence time in half by not discarding the first half to equilibration, but instead starting at the analytical volume and seeing if it drifts away?

Reply:
Thank you! These comments are partially addressed in https://github.com/proteneer/timemachine/pull/433/commits/64520756b9f155edef466049d3f1f75e087895bd by removing the `max_delta_volume` argument, and initializing in a way that should require less equilibration time.

A previous version of this test did initialize at the analytical volume and wait to see if it drifts away, but this led to false passes. Here I've opted to initialize just slightly outside of the relative tolerance, so that the required convergence time is reduced, but the test will still fail if the barostat does nothing.